### ISSUE TITLE
False positive E0308 when trying to access Box with `.0`

### ISSUE THREAD

**Issue by Obstbeeren** [intellij-rust/intellij-rust#10695]
## Environment

* **IntelliJ Rust plugin version:** 0.4.198.5409-231
* **Rust toolchain version:** 1.71.0 (8ede3aae2 2023-07-12) x86_64-unknown-linux-gnu
* **IDE name and version:** PyCharm 2023.1.4 Professional Edition (PY-231.9225.15)
* **Operating system:** Linux 6.4.2-3-manjaro
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.macros.proc.attr

## Problem description
False positive E0308 when accessing Box<String> without dereferencing it.

```
Mismatched types [E0308]
expected `String`, found `Unique<Container>` 
```

## Steps to reproduce
```rust
pub struct Container(pub String);

fn test(x: String) {
    println!("{}", x);
}

fn main() {
    let string = Box::new(Container("test".to_string()));

    // Error: mismatched types [E0308]
    test(string.0);

    let string2 = Box::new(Container("test2".to_string()));
    // No error
    test((*string2).0);
}
```

**Comment by Maksim20023** [comment]
Hi! Thank you for both of your reports.
I forwarded your cases and steps for reproducing it to our developer's team. It will be helpful for us. If you have any further suggestions, please feel free to reach out.
